refactor: async await

vasco-santos opened the pull request:
This PR refactors `libp2p-stardust` and its server to have an async await API, which is compliant with the new [interface-transport](https://github.com/libp2p/js-interfaces/tree/master/src/transport), [interface-connection](https://github.com/libp2p/js-interfaces/tree/master/src/connection) and [interface-peer-discovery](https://github.com/libp2p/js-interfaces/tree/master/src/peer-discovery).

In addition, the stardust server was also completely refactor to work with the new transport.

There is some stuff that I would like to be different in this module, such as what I described in the next comment. However, I think that we should move that into a new issue to be considered later if relevant.

Needs:

- [x]  Update docs
- [ ]  Connection encryption redundancy

Follow up tasks (create issues):

- [x] `interface-transports` tests should be refactored to be used in this module (as well as for circuit relay)
- [x] new protobuf definition?

vasco-santos committed:
refactor: async await

vasco-santos committed:
chore: use libp2p

vasco-santos committed:
chore: not use micro-switch anymore

vasco-santos committed:
chore: use classes instead

vasco-santos commented:
@mkg20001 @jacobheun 

I currently have the register and discovery flows working with the new `js-libp2p` dialer + new transport/discovery interfaces. I am currently working on the dial flow.

Since starting to tackle this, I was thinking on suggesting that we change the naming of the messages in a more meaningful way. However, I wanted to discuss this with both of you after getting an initial version working.

Taking into account the current state for the dial flow, I got to a point when I really want to change the protobuf definition. 

1. Have a single top level messages instead of a bunch of different ones. Considering for example [libp2p-circuit](https://github.com/libp2p/js-libp2p/blob/8bed8f39ff3975850191cec93b7537d1428c8629/src/circuit/protocol/index.js), or [libp2p-daemon](https://github.com/libp2p/js-libp2p-daemon/blob/master/src/protocol/index.js). This way, we should have a Stardust message which includes all the other messages inside, like the daemon example. The main advantage of this consists on being easy to identify the type of the message from the start, which is particularly nice when the server needs to understand if a message from a new stream should go to the register or dial flows, without needing to decode and inspect the content of the message to infer its type.

2. Simplify the messages naming and parameters:

```
JoinInit {random128, peerId} => Register { data, peerId }
JoinChallenge { error, saltEncrypted } => Challenge { data } / Error { msg }
JoinChallengeSolution { solution } => Solution { data }
JoinVerify { error } => Ok { } / Error { msg }

Discovery { ids } => Discovery { ids }
DiscoveryAck { } => Ok { }

DIalRequest { target } => Dial { }
DialResponse { } => Ok { }
```

I will start working on changing the **first** part while I wait for your feedback, as it will ease a lot the implementation of the dial flow. Anyway, I would like feedback from you on both this points before getting this ready to review.

vasco-santos committed:
chore: add functional dial flow

vasco-santos committed:
chore: update tests

vasco-santos committed:
chore: enable browser tests

vasco-santos commented:
FYI, I did not do the protobuf changes above, and I think that I will pass them into an issue to be discussed, instead of being included in this PR.

mkg20001 commented:
My reasoning for not wanting to use a single message is that it doesn't matter if you put in a valid but wrongly typed message, it's still invalid. Except there's more stuff to check for the fields existing. And more code to be executed per parsed message since it needs to parse all the other non-existing messages aswell.
And more code is more bugs. So unless there would be any advantage for the actual code itself, let's just not add complexity for no reason.

Edit: Also the type is explicitly specified by the order, so no checking required.

vasco-santos committed:
chore: update dependencies

vasco-santos committed:
chore: use libp2p utils

mkg20001 commented on README.md:
```diff
@@ -1,184 +1,71 @@
+const { collect } = require('streaming-iterables')
 
-const Stardust = require('libp2p-stardust')
+const addr = multiaddr('/ip4/188.166.203.82/tcp/5892/ws/p2p-stardust')
```
shouldn't we give this a proper dnsaddr? or is this just a tmp addr?

vasco-santos replied:
yeah, after this PR being merged, we will need to deploy the new version of the server, and by then we can update this

mkg20001 replied:
tracking here https://github.com/libp2p/js-libp2p-stardust/issues/15

vasco-santos replied:
Thanks ❤️

mkg20001 commented on src/index.js:
```diff
@@ -1,101 +1,140 @@
+   * @param {Multiaddr} ma
+   * @param {object} options
+   * @param {AbortSignal} options.signal Used to abort dial requests
+   * @returns {Promise<SimplePeer>} Resolves a SimplePeer Webrtc channel
```
webrtc. sounds wrong...

vasco-santos replied:
yes, that's what happen when you copy js docs 😅

mkg20001 commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
+   * @param {Array<multiaddr>} options.addresses
+   * @param {Array<Transport>} options.transports
+   * @param {Array<Multiplexer>} options.muxers
+   * @param {Array<Encryption>} options.encryption
```
The idea was to use stardust over wss:// protocol, making it encrypted already, and dropping the extra secio layer.
Connections between peers are always encrypted end-to-end encrypted and authed by secio, so even plaintext stardust isn't posing a security risk (aside from allowing praying eyes to view metadata)

Since it was planned to be used over wss:// the SECIO layer will only slow things down.

vasco-santos replied:
@mkg20001 yeah, I know that not using the extra encryption layer was one the ideas here. I focused on getting this to work with the new libp2p api first. Now that things are getting shaped, I will understand if we will be able to not use the SECIO layer

mkg20001 replied:
Ok

Possibly it could just use a switch with plaintext, otherwise making a new micro-switch would also be possible should the libp2p switch be either too complicated to run without a swarm or cause significant performance impacts

We should also look out to prevent the user from having to enter an id, even if it's just a placeholder for plaintext-crypto. If not possible to remove from the switch a fake id could always be appended by the listener in that case.

vasco-santos replied:
We should make the libp2p dialer working for all types of use cases, so I think that we should not go on creating a new "switch" to get this and understand what can be done without compromising security. In this way, we cannot use the Plaintext crypto protocol as peers would have to add it to use `Stardust` 🙅‍♂

We can potentially have an [upgrader](https://github.com/libp2p/js-libp2p/blob/master/src/upgrader.js) here, that would basically allow us to upgrade the connection manually instead of using the provided upgrader (this would an improvement over the `micro-switch` idea).

I was discussing this with @jacobheun earlier today and he pointed out an important aspect. Our `websocket` transport is relying on the underlying ws module to verify if the data is encrypted and it is not considering `wss` / `ws`. This creates a risk for this.

We should probably think a bit about the alternatives that we have and understand what will be the best decision.

mkg20001 replied:
> We can potentially have an upgrader here, 

Sounds good

> ws module to verify if the data is encrypted and it is not considering wss / ws. This creates a risk for this.

As I already mentioned earlier the actual p2p connections are encrypted and authed by SECIO

What is possible with using a non-encrypted stardust connection is just DOSing the client with false discovery with a MITM information (this would only work if a MITM would happen and additionally would create at most a minor inconvenience) and the leakage of some metadata (who talks to whom) which would be leaked with regular TCP anyways as the SECIO-handshake is unencrypted.

Stardust is also mainly intended for websites that can't directly dial to peers so the website is most likely using HTTPS which means that we're not _able_ to use unencrypted ws

So unless I'm missing something it appears to be a wrong assumption that this creates a problem.

Though checking if the underlying connection is ws or wss isn't hard as the transport is the one that's dialing so it should be enough to match it the multiaddr in question uses ws or wss

mkg20001 replied:
So if we insist on encryption we could do:
 - if ws: custom upgrader with SECIO (or user-specified) as only crypto
 - if wss: custom upgrader with plaintext as only crypto

The server would use both plaintext and SECIO then

jacobheun replied:
One thing we need to keep in mind is that stardust allows for any `reliable` address, which includes tcp, webrtc, http, etc. We are also not even checking that, just that the stardust multiaddr codec exists in the address. If we remove the potential additional encryption layer, we should be more strict about the addresses we allow. As @mkg20001 mentioned, the general use case is websockets in browser, which will mandate wss. However, I don't see anything in the dialer/listener preventing me from connecting to an insecure tcp stardust server if I discover the address, but as already mentioned it's a low threat attack vector.

We could enforce wss and use plaintext, but the fact that stardust potentially supports any reliable multiaddr is pretty nice. If we want to support broader use cases like `tcp > stardust > websockets` broadly using secio to reduce the complexity might be worth the throughput hit. If we just care about `wss > stardust > websockets`, I think we can enforce that and reasonably use a plaintext upgrader.

vasco-santos replied:
I suggest .that we take this out of the scope of the refactor PR and discuss this on an issue. I think we should be able to provide options to the transport/server that would allow the users to operate with both of those cases. What do you think? Can you create an issue @mkg20001 ?

jacobheun replied:
I think it would be reasonable to take the double encryption hit for now and do a followup release for performance tuning for transports with built in encryption. It's not ideal but mitigates any risk/complexity of multiple reliable multiaddrs and gets the transport usable for folks.

mkg20001 commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
 
-      rpc.writeProto(JoinVerify, {})
+    if (solution.toString('hex') !== solutionClient.toString('hex')) {
+      wrappedStream.writePP({ error: Error.E_INCORRECT_SOLUTION }, JoinVerify)
```
```suggestion
      wrappedStream.writePB({ error: Error.E_INCORRECT_SOLUTION }, JoinVerify)
```

mkg20001 commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
-        return muxed.end()
-      }
+    if (!Buffer.isBuffer(random) || random.length !== 128) {
+      wrappedStream.writePP({ error: Error.E_RAND_LENGTH }, JoinVerify)
```
```suggestion
      wrappedStream.writePB({ error: Error.E_RAND_LENGTH }, JoinVerify)
```

mkg20001 reviewed (changes requested):
Not sold on using an extra SECIO layer, due to e2e-secio and wss being assumed

vasco-santos committed:
chore: minor tweaks

vasco-santos committed:
chore: apply suggestions from code review

Co-Authored-By: Maciej Krüger <mkg20001@gmail.com>

vasco-santos committed:
chore: add upgrader to inbound conns

vasco-santos commented:
I think this is now ready for a first complete review

cc @mkg20001 @jacobheun

vasco-santos committed:
chore: update libp2p-utils

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
 # js-libp2p-stardust
 
-[![](https://img.shields.io/badge/made%20by-mkg20001-blue.svg?style=flat-square)](https://mkg20001.io)
+[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
```
Change this back to mkg.

vasco-santos replied:
yeah, my bad! I copied this from other repo and did not realize

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
 
-[![](https://img.shields.io/badge/made%20by-mkg20001-blue.svg?style=flat-square)](https://mkg20001.io)
+[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
+[![](https://img.shields.io/badge/project-IPFS-blue.svg?style=flat-square)](http://ipfs.io/)
```
Badges should be for libp2p not ipfs

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
-[![Coverage](https://coveralls.io/repos/github/libp2p/js-libp2p-stardust/badge.svg?branch=master)](https://coveralls.io/github/libp2p/js-libp2p-stardust?branch=master)
-[![david-dm](https://david-dm.org/libp2p/js-libp2p-stardust.svg?style=flat-square)](https://david-dm.org/libp2p/js-libp2p-stardust)
+[![Discourse posts](https://img.shields.io/discourse/https/discuss.libp2p.io/posts.svg)](https://discuss.libp2p.io)
+[![Coverage Status](https://coveralls.io/repos/github/libp2p/js-libp2p-websockets/badge.svg?branch=master)](https://coveralls.io/github/libp2p/js-libp2p-websockets?branch=master)
```
Change to stardust repo, not websockets

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
+[![Discourse posts](https://img.shields.io/discourse/https/discuss.libp2p.io/posts.svg)](https://discuss.libp2p.io)
+[![Coverage Status](https://coveralls.io/repos/github/libp2p/js-libp2p-websockets/badge.svg?branch=master)](https://coveralls.io/github/libp2p/js-libp2p-websockets?branch=master)
+[![Travis CI](https://travis-ci.org/libp2p/js-libp2p-stardust.svg?branch=master)](https://travis-ci.org/libp2p/js-libp2p-stardust)
+[![Circle CI](https://circleci.com/gh/libp2p/js-libp2p-stardust.svg?style=svg)](https://circleci.com/gh/libp2p/js-libp2p-stardust)
```
Remove circle ci badge
```suggestion
```

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
+[![Dependency Status](https://david-dm.org/libp2p/js-libp2p-stardust.svg?style=flat-square)](https://david-dm.org/libp2p/js-libp2p-stardust)
+[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)
+![](https://img.shields.io/badge/npm-%3E%3D3.0.0-orange.svg?style=flat-square)
+![](https://img.shields.io/badge/Node.js-%3E%3D4.0.0-orange.svg?style=flat-square)
```
Get rid of the npm and node badges. They're wrong and not necessary.

jacobheun commented on README.md:
```diff
@@ -1,184 +1,75 @@
 
 [![](https://raw.githubusercontent.com/libp2p/interface-transport/master/img/badge.png)](https://github.com/libp2p/interface-transport)
 [![](https://raw.githubusercontent.com/libp2p/interface-connection/master/img/badge.png)](https://github.com/libp2p/interface-connection)
 [![](https://github.com/libp2p/interface-peer-discovery/raw/master/img/badge.png)](https://github.com/libp2p/interface-peer-discovery)
```
I would just remove these badges, the interface repos are out of date.

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
-const EE = require('events').EventEmitter
+const { EventEmitter } = require('events')
+const { AbortError } = require('abortable-iterator')
+// const mafmt = require('mafmt')
```
```suggestion
```

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
+   * @param {Muxer[]} options.muxers - Muxer(s) for microswitch
+   * @param {boolean} options.softFail - Whether to softly fail on listen errors
+   */
+  constructor ({ upgrader, libp2p = {}, id, softFail }) {
```
- No need to pass `id` in, we can get it from `libp2p.peerInfo.id`.
- `softFail` should probably default to `false`

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
 
-    callback = once(callback || noop)
-
     const server = this.connections[getServerForAddress(ma)]
```
`server` can be undefined, which means we're going to throw an error trying to use `server._dial`. If an existing connection is required for dialing, we should throw a meaningful error.

Ideally, we should try to connect to the server if we don't have an existing connection. Right now, I can't dial another peer unless I am already connected to the same server, which is not great.

vasco-santos replied:
I agree with trying to connect to the server if we don't have an existing connection. I will create an issue to add that feature to not create noise to this PR

mkg20001 replied:
The original reason for not wanting to dial a server we aren't connected to was that p2p-stardust wasn't meant to be used like a regular connection and it should instead try any other (non-relay/direct) method

This method should be similarly de-prioritized like p2p-circuit is to avoid that (or rather direct connections should be preferred over relayed ones, so that those only get used as a last resort)

vasco-santos replied:
Created an [libp2p/js-libp2p-stardust#17](https://github.com/libp2p/js-libp2p-stardust/issues/17) to discuss and track this

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
-      if (includes(ma.protoNames(), 'p2p-stardust')) {
+      if (ma.protoNames().includes('p2p-stardust')) {
         return true
       }
```
The last line in this function is commented out:
> // return mafmt.Stardust.matches(ma)

while not strictly required, we should `return false` as the default and just remove that commented out code.

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
       options = {}
     }
 
     handler = handler || noop
```
The noop is unnecessary, we can just have the listener check before calling `handler && handler(...)`. I'd remove it and it's declaration at the top of the file.

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+log.error = debug('libp2p:stardust:listener:error')
 
-const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
+const EventEmitter = require('events')
```
We should be consistent with the imports, `src/index.js` is importing this way:
```suggestion
const { EventEmitter } = require('events')
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-const log = debug('libp2p:stardust:client')
+const Wrap = require('it-pb-rpc')
+const { int32BEDecode, int32BEEncode } = require('it-length-prefixed')
+const { JoinInit, JoinChallenge, JoinChallengeSolution, JoinVerify, Discovery, DiscoveryAck, DialRequest, DialResponse, ErrorTranslations } = require('./proto')
```
nit: add these on their own lines to more easily see what's being imported.

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+   * @param {Object} properties - properties for the listener
+   * @param {function (Connection)} properties.handler - New connection handler
+   * @param {Upgrader} properties.upgrader - Connection upgrader
+   * @param {Upgrader} properties.client - Stardust client reference
```
```suggestion
   * @param {Stardust} properties.client - Stardust client reference
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+  /**
+   * @constructor
+   * @param {Object} properties - properties for the listener
+   * @param {function (Connection)} properties.handler - New connection handler
```
handler should be optional
```suggestion
   * @param {function (Connection)} [properties.handler] - New connection handler
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+   * @param {function (Connection)} properties.handler - New connection handler
+   * @param {Upgrader} properties.upgrader - Connection upgrader
+   * @param {Upgrader} properties.client - Stardust client reference
+   * @param {Object} properties.options - options for the listener
```
Should be optional
```suggestion
   * @param {Object} [properties.options] - options for the listener
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-    this._listen(ma).then(() => {
+  /**
+   * Listen on a multiaddr (from stardust server)
+   * @param {multiaddr} ma multiaddr to listen on
```
```suggestion
   * @param {Multiaddr} ma multiaddr to listen on
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-      } else {
-        this.emit('error', err)
-        callback(err)
+        this.emit('listening')
```
This is not accurate. We failed to listen, we shouldn't emit a listening event regardless of if failures are allowed.

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-    cb(null, this.address ? [this.address] : [])
+  /**
+   * Get listener addresses
+   * @returns {Array<multiaddr>}
```
```suggestion
   * @returns {Array<Multiaddr>}
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-    const addrBase = this.address.decapsulate('p2p-stardust')
+  /**
+   * Connect to a stardust server.
+   * @param {multiaddr} addr address of the stardust server
```
```suggestion
   * @param {Multiaddr} addr address of the stardust server
```

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
-const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
+const EventEmitter = require('events')
 
 const crypto = require('crypto')
```
We could potentially just use `libp2p-crypto` to reduce bundle size. The `sha5` function would need to be updated though. This could be done in a followup PR though to isolate that change.

vasco-santos replied:
I will create an issue to track that!

vasco-santos replied:
#18

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+   * Read discovery messages
+   * @private
+   */
+  async _onDiscovery () {
```
nit: `_onDiscovery` is a misnomer, we're not reacting to anything, we're actively trying to read peers. Might be good to change this to something like `_discoverPeers`

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+    log('reading discovery')
+    message.ids
+      .map(id => {
+        const pi = new PeerInfo(new PeerId(id))
```
This can throw, we need to catch the error.

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+        return pi
+      })
+      .filter(Boolean)
+      .forEach(pi => this.client.discovery.emit('peer', pi))
```
This chain is iterating over the ids 3 times (map, filter, forEach). I would just turn this into a `for...of` loop and move all the logic into that.

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
+      .filter(Boolean)
+      .forEach(pi => this.client.discovery.emit('peer', pi))
+
+    setTimeout(() => this._onDiscovery(), 100) // cooldown
```
100ms seems unnecessarily frequent. We will immediately discover on the initial listen, so a bigger timeout would be ideal, maybe even 10s. We should also make this a configurable interval.

jacobheun replied:
After reading through the server code I think this discovery logic might be better as an async iterable where we continuously read from the stream. Since the server is broadcasting to our stream there's no reason to continually try to read. If we get disconnected we can execute the reconnect logic and continue reading. This would remove the need for any timeout logic.

vasco-santos replied:
I agree with the 100ms being too frequent. I was trying to avoid changing these things in the refactor PR.

I think that adding the option is a good idea though. So, if I am changing this line, I am happy to also change the default

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,234 @@
 
     const id = addr.getPeerId()
-    const _id = ID.createFromB58String(id)._id
+    const _id = PeerId.createFromB58String(id)._id
```
We can use the getter to avoid using the private property `_id`
```suggestion
    const _id = PeerId.createFromB58String(id).id
```

jacobheun commented on src/index.js:
```diff
@@ -1,98 +1,136 @@
     this.softFail = softFail
 
-    this.discovery = new EE()
+    this.connections = {}
```
It might be good to change this to `this.listeners` for clarity, as it's a map of Listener objects, not connections.

vasco-santos replied:
yeah, good call

jacobheun commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
-  const targetB58 = new ID(target).toB58String()
-
-  log('dial from %s to %s', id.toB58String(), targetB58)
+const Libp2p = require('libp2p')
```
This is a dev dependency but is being used in src code.

jacobheun commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
+const Wrap = require('it-pb-rpc')
+const pipe = require('it-pipe')
+const { int32BEDecode, int32BEEncode } = require('it-length-prefixed')
+const { JoinInit, JoinChallenge, JoinChallengeSolution, JoinVerify, Discovery, DiscoveryAck, DialRequest, DialResponse, Error } = require('../proto')
```
nit: Split on newlines

jacobheun commented on src/server/bin.js:
```diff
@@ -6,23 +6,22 @@
 
-const server = new Server({ addresses })
-server.start().then(() => {
-  server.switch.addresses.map(String).forEach(addr => console.log('Listening on %s', addr))
```
Are we losing the logging of the addresses?

jacobheun commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
+   */
+  update () {
+    log('updating cached data')
+    this.networkArray = Object.keys(this.network).map(b58 => this.network[b58])
```
just use `Object.values`
```suggestion
    this.networkArray = Object.values(this.network)
```

jacobheun commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
+    log('broadcasting discovery to %o client(s)', this.networkArray.length)
+    this.networkArray.forEach(client => {
+      client.wrappedStream.writePB(this._cachedDiscovery, Discovery)
+    })
```
Out of scope for this PR, but it might be good to keep a timestamp of last broadcast per peer so we only need to send them peers they haven't seen.

vasco-santos replied:
yeah 👍 , I will create one issue

mkg20001 replied:
The current behavior leans on that of the wrtc discovery module, which is why that is.

Also I optimized things so it does only one serialization of the entire list, but this means that we can't send different lists to different peers.

But I don't think that scale will be reached, so just sending different lists to different peers with just the newly joined ones should do it.

(There's a catch: If the peer count is zero a zero-length list (0x01) needs to be sent, otherwise the data that gets sent is zero-length - That is currently implemented)

vasco-santos replied:
Created #19 to further discussion and track this improvement

jacobheun commented on src/server/index.js:
```diff
@@ -1,188 +1,228 @@
+      wrappedStream
     }
+
+    checkAckLoop(wrappedStream, () => this.removeFromNetwork(client))
```
What is the value of the ack check? Right now after we broadcast the peers should respond with their ack messages. If we don't we're not disconnecting them, we just stop broadcasting to them. It seems like we should just pay more attention to the original stream. 

When a peer connects we perform a handshake. After the handshake we start streaming discovered peers to them. If the peer is not acting as a listener (they're just dialing through us), they can close the stream to not receive discovery events. Anytime a stream is closed we can remove that peer from our streams to pipe known peers to. This removes any need for an ack check. The ack check does add a "ping" mechansim to ensure peers are responding to us, but I don't know that that's necessary.

mkg20001 replied:
TCP can leave dead connections sometimes for a really long time (especially if no data gets written by the still-alive side) if the other side vanishes (for example someone closing their laptop). So those closed-laptop zombie peers could be left around for minutes, unnecessarily polluting other peers discovery events (unless we're sending something to it, in which case the connection gets dropped quicker).

Also NATs tend to close connections that aren't sending data for some time

So if we dialed a peer through a relay with disabled discovery after we disconnect from that peer and dial another through the same realay after some time we'd end up with TCP errors because the NAT remove the connection pair from the table (so both sides still think the connection is open, but it's not)

jacobheun commented on test/compliance.spec.js:
```diff
@@ -0,0 +1,51 @@
+const Stardust = require('../src')
+const { createPeer } = require('./utils')
+
+describe.skip('interface-transport compliance', () => {
```
Why are we skipping, add a comment if it's needed?

vasco-santos replied:
I have the reason for it on the PR description. But I will create an issue and add its url here

vasco-santos replied:
#21

jacobheun reviewed (changes requested):
In addition to the comments I've left, I haven't reviewed the tests in full but I think we should add some more verbose testing around discovery of multiple peers. My main concern is around discovery and ensuring that we are properly closing streams. If discovery stays the way it is, we should ensure that if the server doesn't receive an `ACK` in time and removes the peer from tracking, that we're not leaving that stream open. We need more scenario testing around this, especially failure states.

vasco-santos committed:
chore: address review

vasco-santos committed:
chore: added tests for discovery and server

vasco-santos commented:
@jacobheun I think that I addressed all your feedback. 

Regarding tests for discovery / streams, thanks for raising this. I found a flow with problems, as we were not closing the streams on `listener.close`. I added a few tests for discovery, and also one complete discovery flow test for the server. Let me know your thoughts

vasco-santos committed:
chore: change author to leadMaintainer to be fetched via the package-table tool

jacobheun commented on README.md:
```diff
@@ -1,184 +1,68 @@
-[![david-dm](https://david-dm.org/libp2p/js-libp2p-stardust.svg?style=flat-square)](https://david-dm.org/libp2p/js-libp2p-stardust)
+[![Discourse posts](https://img.shields.io/discourse/https/discuss.libp2p.io/posts.svg)](https://discuss.libp2p.io)
+[![Coverage Status](https://coveralls.io/repos/github/libp2p/js-libp2p-stardust/badge.svg?branch=master)](https://coveralls.io/github/libp2p/js-libp2p-stardust?branch=master)
+[![Travis CI](https://travis-ci.org/libp2p/js-libp2p-stardust.svg?branch=master)](https://travis-ci.org/libp2p/js-libp2p-stardust)
```
use travis.com instead of .org
```suggestion
[![Travis CI](https://travis-ci.com/libp2p/js-libp2p-stardust.svg?branch=master)](https://travis-ci.com/libp2p/js-libp2p-stardust)
```

jacobheun commented on test/compliance.spec.js:
```diff
@@ -0,0 +1,51 @@
+const Stardust = require('../src')
+const { createPeer } = require('./utils')
+
+describe.skip('interface-transport compliance', () => {
```
```suggestion
// Skipped due to https://github.com/libp2p/js-libp2p-stardust/issues/21
describe.skip('interface-transport compliance', () => {
```

jacobheun commented on test/dial.spec.js:
```diff
@@ -1,49 +1,70 @@
-        conn,
-        pull.collect(cb)
-      )
+  it('dial on IPv4, check promise', async function () {
```
Test names should be clearer about what's being tested
```suggestion
  it('dial on IPv4 should return a valid connection', async function () {
```

jacobheun commented on test/dial.spec.js:
```diff
@@ -1,49 +1,70 @@
-      require('assert')(String(pi.multiaddrs.toArray()[0]) === String(SERVER_URL) + String(clients[1].addr))
-      done()
-    })
+  it('dial offline / non-exist()ent node on IPv4, check promise rejected', async function () {
```
```suggestion
  it('dialing a nonexistent peer should fail (IPv4)', async function () {
```

jacobheun commented on test/discovery.spec.js:
```diff
@@ -0,0 +1,128 @@
+      await deferred.promise
+    })
+
+    it('should close connection with server on closing listener after discovery trigger', async function () {
```
```suggestion
    it('closing the listener should close the underlying connection to the stardust server with open discovery streams', async function () {
```

jacobheun commented on test/discovery.spec.js:
```diff
@@ -0,0 +1,128 @@
+    })
+
+    it('discovers all the nodes registered in the server', async function () {
+      this.timeout(20e3)
```
There are several large timeouts in this test suite. Can we lower the discovery interval for the tests to avoid this?

jacobheun commented on test/discovery.spec.js:
```diff
@@ -0,0 +1,128 @@
+        }
+      })
+
+      await deferred.promise
```
`p-wait-for` could simplify this quite a bit. It would probably also be good to validate the peers discovered are the ones we expect.

```js
const discovered = []
clients[3].discovery.on('peer', (peer) => discovered.push(peer))

await pWaitFor(() => discovered.length === 3)
// Do peer validation...
```

jacobheun commented on test/instance.spec.js:
```diff
@@ -0,0 +1,30 @@
+    [libp2p] = await createPeer()
+  })
+
+  it('should be creatable', () => {
```
Maybe add a test for the missing upgrader/libp2p?

jacobheun commented on test/listen.spec.js:
```diff
@@ -0,0 +1,64 @@
+    stardust = new Stardust({ upgrader: mockUpgrader, libp2p })
+  })
+
+  it('listen, check for promise', async () => {
```
improve description about what's being tested
```suggestion
  it('should be able to listen on a valid server', async () => {
```

jacobheun commented on test/server.node.js:
```diff
@@ -0,0 +1,79 @@
+  })
+
+  afterEach(async () => {
+    sinon.reset()
```
Do you want `sinon.restore()`? I believe reset will keep the fakes in place but just reset them. `restore` will remove all fakes and restore the originals.

jacobheun commented on test/server.node.js:
```diff
@@ -0,0 +1,79 @@
+    expect(spyRemoveFromNetwork.called).to.eql(false)
+
+    // Trigger a disconnect from the client
+    await listener.close()
```
This is just disconnecting the client directly. We should have a test for what happens when a client is connected, but fails to respond to the ack. I think this should result in the server untracking the peer **and** disconnecting them.

jacobheun reviewed (changes requested):
Added some suggestions/improvements focused primarily on tests.

vasco-santos committed:
chore: apply suggestions from code review

Co-Authored-By: Jacob Heun <jacobheun@gmail.com>

vasco-santos committed:
chore: address latest review comments on tests

vasco-santos commented:
@jacobheun made a new pass on the tests after your comment. Found a fixed a bug with that 🎉

vasco-santos committed:
chore: add bin env to server

jacobheun commented on src/listener.js:
```diff
@@ -1,183 +1,266 @@
+
+    // close stream and connection with the server
+    const wrappedStream = this.wrappedStream.unwrap()
+    wrappedStream.sink([])
```
This should be unnecessary. The call below to `this.serverConnection.close()` is terminating the entire connection, so `this.wrappedStream`, which is a stream of that connection should automatically terminate.

Also, just a note, `wrappedStream.sink([])` **ONLY** closes the write side of the stream, the read would still be open. `source` and `sink` must both be closed. We should document this better in the libp2p and connection interface docs, with examples of how to properly close streams to help people avoid halfopen streams.

vasco-santos replied:
> We should document this better in the libp2p and connection interface docs, with examples of how to properly close streams to help people avoid halfopen streams.

Yeah, it makes sense

jacobheun commented on test/discovery.spec.js:
```diff
@@ -0,0 +1,123 @@
+      }, conn => pipe(conn, conn)))
+      await Promise.all(listeners.map(listener => listener.listen(SERVER_URL)))
+
+      await pWaitFor(() => discovered.length >= 3)
```
We should validate the discovered peer list after the wait for. Also, it might be good to make `discovered` a `key:peerId` Map to avoid discovered being larger than 3. We should not find more than 3 peers, the excess is likely due to "rediscovering" peers we already know.

jacobheun reviewed (commented):
Just two minor things. I think once we merged this it would be good to do an RC and test it out in js-ipfs before doing a final release.

vasco-santos committed:
chore: remove redundant finish of sink

mkg20001 commented on README.md:
````diff
@@ -1,184 +1,68 @@
 
-```bash
-> npm install libp2p-stardust
+console.log(`Value: ${values.toString()}`)
````
values is an array, so I think that won't work

Also this example doesn't fully work, it would be great if the example in the readme would fully work on it's own

vasco-santos replied:
I agree! I was thinking on getting it updated once we have a server deployed, so that we can point to it. 

I will  change the example so that it is "workable" and once we deploy it, we can put the correct server address :) I will create an issue for tracking this!

mkg20001 commented on src/index.js:
```diff
@@ -1,102 +1,144 @@
         return true
       }
 
-      // return mafmt.Stardust.matches(ma)
```
I suppose it would be better if we fix & update mafmt to match the addresses correctly. I only commented this out and added the other snippet as a quickfix some time ago.

vasco-santos replied:
Yes, I initially was thinking the same. 

We should postpone this to after deciding wether stardust should be usable on reliable addresses, or just ws/wss, or even just wss. I forgot to open an issue to discuss this, but I will open it now and we can make this happen in a followup PR after a decision is made.

mkg20001 replied:
Ok, sounds good. My opinion is that (since the security concern is mainly DOS) it's not wrong to allow it, yet there should be a way we make it clear to the user that it might be an issue that way.

Or stardust could check if the address has an encryption layer (for ex wss) and add one if necessary (using any of the libp2p crypto layers). We could accomplish this by adding secio+plaintext to the server and then using either a plaintext or secio upgrader on the client.

vasco-santos replied:
can you post that on #22 ? It will be easier to discuss from there

mkg20001 commented on src/listener.js:
```diff
@@ -1,183 +1,263 @@
 
-    muxed.on('stream', this.handler)
-    this._readDiscovery()
+    this._timeoutId = setTimeout(() => this._discoverPeers(), this.options.discoveryInterval || 10000) // cooldown
```
This will queue up messages if the server's discovery interval is higher. We can either exchange the interval via the join success message or just directly read (though the second one is better since connection lags can get this out of hand and break the ACK mechanism - thus disconnecting the peer)

vasco-santos replied:
Yes, I agree that we can remove the timeout and just block on the stream read

mkg20001 commented on src/server/bin.js:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env node
```
permissions should be 755 (are 644 right now, which is missing x bit)

vasco-santos replied:
Good catch!

mkg20001 commented on src/server/index.js:
```diff
@@ -1,188 +1,244 @@
+    const { solution: solutionClient } = await wrappedStream.readPB(JoinChallengeSolution)
 
-      rpc.writeProto(JoinVerify, {})
+    if (solution.toString('hex') !== solutionClient.toString('hex')) {
```
If anyone knows a better method to directly compare buffers it would be great, this is more of a hack from me

vasco-santos replied:
Any reason for not using https://nodejs.org/api/buffer.html#buffer_buf_compare_target_targetstart_targetend_sourcestart_sourceend ?

I tried to keep the logic the same to not make the refactor noisy

mkg20001 replied:
Me not knowing this exists :P

vasco-santos replied:
ahah, I will change it then :)

mkg20001 reviewed (changes requested):
Nitpicks mostly, otherwise good to go

vasco-santos committed:
chore: address review

vasco-santos committed:
chore: address review

vasco-santos committed:
fix: use correct listen addr

vasco-santos committed:
chore: merge

vasco-santos commented:
@mkg20001 and @jacobheun made a few changes to this. Listen should take into account `multiaddr` like `/ip4/0.0.0.0/tcp/5892/ws/p2p-stardust/p2p/${stardustServerId}/p2p/${ownId}`.

Also tested this with browser examples of `js-libp2p` and `js-ipfs`

mkg20001 commented on src/server/index.js:
```diff
@@ -143,7 +143,7 @@ class Server {
     const { solution: solutionClient } = await wrappedStream.readPB(JoinChallengeSolution)
 
-    if (solution.toString('hex') !== solutionClient.toString('hex')) {
+    if (solution.compare(solutionClient) !== 0) {
```
```suggestion
    if (solution.compare(solutionClient)) { // if not equal will return non-zero value
```

mkg20001 commented on src/utils.js:
```diff
@@ -0,0 +1,18 @@
+    }
+  })
+
+  // Get firs id
```
```suggestion
  // Get first id
```

mkg20001 reviewed (approved):
Last batch of nitpicks to apply, after that LGTM

vasco-santos committed:
chore: apply suggestions from code review

Co-Authored-By: Maciej Krüger <mkg20001@gmail.com>

vasco-santos commented:
As @jacobheun suggested, we should probably release this as a `rc` while the examples are reviewed. What do you think @mkg20001 ?

jacobheun reviewed (approved):
LGTM 🚀

mkg20001 commented:
@vasco-santos ok

Since I'm lead I should merge this now? (Haven't done a merge in eons)

vasco-santos commented:
> Since I'm lead I should merge this now? (Haven't done a merge in eons)

Yeah, go for it 🚢

For releasing, you can have a look at [ipfs/aegir/#releasing](https://github.com/ipfs/aegir/#releasing). There is an example for the pre release! 🙂

vasco-santos commented:
You should just need to run:

```sh
npm run release -- --type prerelease --preid rc --dist-tag next
```
